File: 2019/day-1/sol.rb
# setup {{{
aoc_input = File.read('input')

def test(fname, samples)
  f = method(fname)
  puts "TESTING #{fname}"
  samples.each_with_index do |sample, i|
    print " test #{i}: "
    res = sample.last
    r = f.call(*sample[0..-2])
    if r != res
      puts "FAIL "
      puts "   expected"
      puts "     #{res}"
      puts "   but got"
      puts "     #{r}"
    else
      puts "ok"
    end
  end
end
# }}}

def process_input(input)
  input.strip.split.map(&:strip) .map(&:to_i)
end

def fuel(n)
  (n / 3) - 2
end

def fuelr(n)
  f = fuel(n)
  f <= 0 ? 0 : f + fuelr(f)
end

def sol1(input)
  xs = process_input(input)
  xs.map {|n| fuel(n)} .sum
end

def sol2(input)
  xs = process_input(input)
  xs.map {|n| fuelr(n)} .sum
end

# def test(*_) return 'uncomment to disable tests'; end

test :fuel, [
  [12, 2],
  [1969, 654],
]
p sol1 aoc_input
test :fuelr, [
  [1969, 966],
  [100756, 50346],
]
p sol2 aoc_input

File: 2019/day-1/sol.js
// setup {{{
let aocInput;
{
  const fs = require('fs');
  aocInput = fs.readFileSync('input', 'utf-8');
}
// }}}

const processInput = input => input.trim().split('\n').map(Number);

const fuel = mass => Math.floor(mass / 3) - 2;
const fuel2 = mass => fuel(mass) + fuel2r(fuel(mass));
const fuel2r = m => (m > 6 ? fuel(m) + fuel2r(fuel(m)) : 0);

const sol = input => processInput(input).reduce((acc, m) => acc + fuel(m), 0);

const sol2 = (input) => processInput(input).reduce((acc, m) => acc + fuel2(m), 0);

const test = (fn) => {
  const samples = [
    [['12'], 2],
    [['14'], 2],
    [['1969'], 654],
    [['100756'], 33583],
  ];

  // {{{ test body
  for (const i in samples) {
    const [args, expectedResult] = samples[i];
    const actualResult = fn(...args);

    const strExpected = JSON.stringify(expectedResult);
    const strActual = JSON.stringify(actualResult);

    if (strExpected === strActual) {
      console.log(`test ${i}: ok (${strActual})`);
    } else {
      console.log(`test ${i}: FAILED : expected v`);
      console.group();
      console.group();
      console.log(expectedResult);
      console.groupEnd();
      console.log('instead got v');
      console.group();
      console.log(actualResult);
      console.groupEnd();
      console.groupEnd();
    }
  }
  // }}}
};

const test2 = (fn) => {
  const samples = [
    [['12'], 2],
    [['14'], 2],
    [['1969'], 966],
    [['100756'], 50346],
  ];

  // {{{ test body
  for (const i in samples) {
    const [args, expectedResult] = samples[i];
    const actualResult = fn(...args);

    const strExpected = JSON.stringify(expectedResult);
    const strActual = JSON.stringify(actualResult);

    if (strExpected === strActual) {
      console.log(`test ${i}: ok (${strActual})`);
    } else {
      console.log(`test ${i}: FAILED : expected v`);
      console.group();
      console.group();
      console.log(expectedResult);
      console.groupEnd();
      console.log('instead got v');
      console.group();
      console.log(actualResult);
      console.groupEnd();
      console.groupEnd();
    }
  }
  // }}}
};

// test(sol)


console.log(sol(aocInput));

//test2(sol2)

console.log(sol2(aocInput))
